dll/win32/shell32/changenotify.cpp

```diff
@@ -0,0 +1,632 @@
+        SendMessageW(hwndWorker, CN_DELIVER_NOTIFICATION, (WPARAM)hTicket, pid);
+    else
+        SendNotifyMessageW(hwndWorker, CN_DELIVER_NOTIFICATION, (WPARAM)hTicket, pid);
+}
+
+/*************************************************************************
+ * SHChangeNotifyRegister           [SHELL32.2]
+ */
+EXTERN_C ULONG WINAPI
+SHChangeNotifyRegister(HWND hwnd, INT fSources, LONG wEventMask, UINT uMsg,
+                       INT cItems, SHChangeNotifyEntry *lpItems)
+{
+    HWND hwndWorker, hwndOldWorker = NULL;
+    HANDLE hRegEntry;
+    INT iItem;
+    ULONG nRegID = INVALID_REG_ID;
+    DWORD dwOwnerPID;
+    LPREGENTRY pShare;
+
+    TRACE("(%p,0x%08x,0x%08x,0x%08x,%d,%p)\n",
+          hwnd, fSources, wEventMask, uMsg, cItems, lpItems);
+
+    // sanity check
+    if (wEventMask == 0 || cItems <= 0 || cItems > 0x7FFF || lpItems == NULL ||
+        hwnd == NULL || !IsWindow(hwnd))
+    {
+        return INVALID_REG_ID;
+    }
+
+    // request the new delivery worker window
+    hwndWorker = GetNotificationServer(TRUE);
+    if (hwndWorker == NULL)
+        return INVALID_REG_ID;
+
+    // disable new delivery method in specific condition
+    if ((fSources & SHCNRF_RecursiveInterrupt) != 0 &&
+        (fSources & SHCNRF_InterruptLevel) == 0)
+    {
+        fSources &= ~SHCNRF_NewDelivery;
+    }
+
+    // if it is old delivery method, then create the old delivery worker window
+    if ((fSources & SHCNRF_NewDelivery) == 0)
+    {
+        hwndOldWorker = hwnd = CreateNotificationBroker(hwnd, uMsg);
+        uMsg = WM_BROKER_NOTIFICATION;
+    }
+
+    // The owner PID is the process ID of the new delivery worker.
+    GetWindowThreadProcessId(hwndWorker, &dwOwnerPID);
+
+    EnterCriticalSection(&SHELL32_ChangenotifyCS);
+    for (iItem = 0; iItem < cItems; ++iItem)
+    {
+        // create a registration entry
+        hRegEntry = CreateRegistrationParam(nRegID, hwnd, uMsg, fSources, wEventMask,
+                                            lpItems[iItem].fRecursive, lpItems[iItem].pidl,
+                                            dwOwnerPID, hwndOldWorker);
+        if (hRegEntry)
+        {
+            TRACE("CN_REGISTER: hwnd:%p, hRegEntry:%p, pid:0x%lx\n",
+                  hwndWorker, hRegEntry, dwOwnerPID);
+
+            // send CN_REGISTER to new delivery worker
+            SendMessageW(hwndWorker, CN_REGISTER, (WPARAM)hRegEntry, dwOwnerPID);
+
+            // update nRegID
+            pShare = (LPREGENTRY)SHLockSharedEx(hRegEntry, dwOwnerPID, FALSE);
+            if (pShare)
+            {
+                nRegID = pShare->nRegID;
+                SHUnlockShared(pShare);
+            }
+
+            // free registration entry
+            SHFreeShared(hRegEntry, dwOwnerPID);
+        }
+
+        // if failed, then destroy the old worker.
+        if (nRegID == INVALID_REG_ID && (fSources & SHCNRF_NewDelivery) == 0)
+        {
+            ERR("Old Delivery is failed\n");
+            DestroyWindow(hwndOldWorker);
```

Comment:
One last question, what does this destroy? Because I get the feeling that it could destroy a window that doesn't belong to us.